pinot-spi/src/main/java/org/apache/pinot/spi/data/Schema.java

```diff
@@ -673,7 +679,15 @@ public boolean isBackwardCompatibleWith(Schema oldSchema) {
       FieldSpec oldSchemaFieldSpec = entry.getValue();
       FieldSpec fieldSpec = getFieldSpecFor(oldSchemaColumnName);
       if (!fieldSpec.equals(oldSchemaFieldSpec)) {
-        return false;
+        if (fieldSpec.getDataType() == DataType.BOOLEAN) {
+          // For BOOLEAN data type, replace with STRING and compare again to keep backward-compatibility
+          fieldSpec.setDataType(DataType.STRING);
+          if (!fieldSpec.equals(oldSchemaFieldSpec)) {
+            return false;
+          }
+        } else {
+          return false;
+        }
       }
```

Comment:
Would suggest simplifying this code block (three nested if statements and three returns) to:
```
if (fieldSpec.getDataType() == DataType.BOOLEAN && oldSchemaFieldSpec.getDataType() == DataType.STRING) {
  // Replace BOOLEAN with STRING if existing FieldSpec is STRING.
  fieldSpec.setDataType(DataType.STRING);
}

if (fieldSpec.equals(oldSchemaFieldSpec)) {
  return false;
}
```
otherwise, LGTM.

Reply:
Moved the update logic as Subbu suggested and kept this part the same